vpr/src/device/rr_graph_obj.h

```diff
@@ -0,0 +1,417 @@
+ *
+ * Misc.
+ * 1. dirty_ : FIXME need more inputs 
+ *
+ * Fast look-up
+ * 1. node_lookup_: a fast look-up to quickly find a node in the RRGraph by its type, coordinator and ptc_num
+ *
+ ***********************************************************************/
+#ifndef RR_GRAPH_OBJ_H
+#define RR_GRAPH_OBJ_H
+
+/*
+ * Notes in include header files in a head file 
+ * Only include the neccessary header files 
+ * that is required by the data types in the function/class declarations!
+ */
+/* Header files should be included in a sequence */
+/* Standard header files required go first */
+#include <limits>
+#include <vector>
+
+/* EXTERNAL library header files go second*/
+#include "vtr_vector.h"
+#include "vtr_range.h"
+#include "vtr_geometry.h"
+#include "arch_types.h"
+
+/* VPR header files go second*/
+#include "vpr_types.h"
+#include "rr_graph_fwd.h"
+
+class RRGraph {
+  public: //Types
+    typedef vtr::vector<RRNodeId, RRNodeId>::const_iterator node_iterator;
+    typedef vtr::vector<RREdgeId, RREdgeId>::const_iterator edge_iterator;
+    typedef vtr::vector<RRSwitchId, RRSwitchId>::const_iterator switch_iterator;
+    typedef vtr::vector<RRSegmentId, RRSegmentId>::const_iterator segment_iterator;
+
+    typedef vtr::Range<node_iterator> node_range;
+    typedef vtr::Range<edge_iterator> edge_range;
+    typedef vtr::Range<switch_iterator> switch_range;
+    typedef vtr::Range<segment_iterator> segment_range;
+
+  public: //Accessors
+    //Aggregates
+    node_range nodes() const;
+    edge_range edges() const;
+    switch_range switches() const;
+    segment_range segments() const;
+
+    //Node attributes
+    size_t node_index(RRNodeId node) const; /* TODO: deprecate this accessor as outside functions should use RRNodeId */
+    t_rr_type node_type(RRNodeId node) const;
+    const char* node_type_string(RRNodeId node) const;
+
+    short node_xlow(RRNodeId node) const;
+    short node_ylow(RRNodeId node) const;
+    short node_xhigh(RRNodeId node) const;
+    short node_yhigh(RRNodeId node) const;
+    short node_length(RRNodeId node) const;
+    vtr::Rect<short> node_bounding_box(RRNodeId node) const;
+    /* Node starting and ending points */
+    vtr::Point<size_t> node_start_coordinator(RRNodeId node) const;
+    vtr::Point<size_t> node_end_coordinator(RRNodeId node) const;
+
+    short node_capacity(RRNodeId node) const;
+    short node_fan_in(RRNodeId node) const;
+    short node_fan_out(RRNodeId node) const;
+
+    short node_ptc_num(RRNodeId node) const;
+    short node_pin_num(RRNodeId node) const;
+    short node_track_num(RRNodeId node) const;
+    short node_class_num(RRNodeId node) const;
+
+    short node_cost_index(RRNodeId node) const;
+    e_direction node_direction(RRNodeId node) const;
+    const char* node_direction_string(RRNodeId node) const;
+    e_side node_side(RRNodeId node) const;
+    const char* node_side_string(RRNodeId node) const;
```

Comment:
Should these convenience methods really be part of the RRGraph object?  Shouldn't we just have a free function e_side -> const char *?

Reply:
You can use free functions. I prefer convenience methods as it can short your codes. 
Think about in many scenarios, you need to print some information. Without the functions, you will need two lines of codes.

Reply:
> you will need two lines of codes.

This is incorrect?  

Consider the following case: 

If the underlying code needed the enum to being with (which is very likely), then both `node_side` and `node_side_string` will be called, which is actually worse then calling `node_side` and `side_to_string`.  There is no computation advantage to having the string accessor.  

If the code only needs the string, node_side + side_to_string is equivalent.  This can still be done in one statement.

So there is no actual reason to have a direct string accessor on the rr graph object at all.

Reply:
Make sense. The functions are removed.